esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6104,289 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+    MemoryBuffer& layoutBin, Owned<IFileDescriptor> &fileDesc, IUserDescriptor *udesc,
+    IConstDFUFileAccessRequestBase &req, IEspDFUFileCreateResponse &resp)
+{
+    IEspDFUFileAccessInfo &accessInfo = resp.updateAccessInfo();
+    IArrayOf<IEspDFUFilePart> dfuParts;
+    getFilePartsInfo(context, fileDesc, dfuParts);
+    accessInfo.setNumParts(fileDesc->numParts());
+    accessInfo.setFileParts(dfuParts);
+
+    StringBuffer metaInfo, expiryTime;
+    getFileMeta(metaInfo, fileName, fileDesc, udesc, expiryTime, jobId, req);
+    accessInfo.setMetaInfoBlob(metaInfo);
+    accessInfo.setExpiryTime(expiryTime.str());
+
+    if (layoutBin.length() == 0)
+    {
+        if (!req.getReturnJsonTypeInfo() && !req.getReturnBinTypeInfo())
+            return;
+        resp.setWarning("RecordTypeInfoBin or RecordTypeInfoJson requested, but no type info found from ECLRecordDefinition");
+    }
+
+    if (req.getReturnJsonTypeInfo())
+    {
+        StringBuffer jsonLayout;
+        Owned<IRtlFieldTypeDeserializer> deserializer(createRtlFieldTypeDeserializer());
+        const RtlTypeInfo *typeInfo = deserializer->deserialize(layoutBin);
+        dumpTypeInfo(jsonLayout, typeInfo);
+        if (jsonLayout.length())
+            accessInfo.setRecordTypeInfoJson(jsonLayout.str());
+    }
+    if (req.getReturnBinTypeInfo())
+    {
+        MemoryBuffer binLayout;
+        layoutBin.swapWith(binLayout);
+        if (binLayout.length())
+            accessInfo.setRecordTypeInfoBin(binLayout);
+    }
+}
+
+const char DFUFileIdSeparator = ':';
+
+bool CWsDfuEx::onDFUFileCreate(IEspContext &context, IEspDFUFileCreateRequest &req, IEspDFUFileCreateResponse &resp)
+{
+    try
+    {
+        IConstDFUFileAccessRequestBase &requests = req.getRequestBase();
+        SecAccessFlags accessType = translateToSecAccessFlags(requests.getAccessType());
+        if (SecAccess_None == accessType)
+        {
+            context.setAuthStatus(AUTH_STATUS_NOACCESS);
+            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate - Permission denied.");
+        }
+        context.ensureFeatureAccess(FEATURE_URL, accessType, ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate: Permission denied.");
+
+        const char *fileName = requests.getName();
+        const char *clusterName = requests.getCluster();
+        const char *recordDefinition = req.getECLRecordDefinition();
+        if (isEmptyString(fileName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Name defined.");
+        if (isEmptyString(clusterName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Cluster defined.");
+        if (isEmptyString(recordDefinition))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No ECLRecordDefinition defined.");
+
+        StringBuffer groupName;
+        if (!createWsDFUCreateGroupName(req.getPartLocations(), groupName))
```

Comment:
I think we should make partLocations in the request optional.
If not supplied it uses the group from the Cluster supplied.

Reply:
@rpastrana is this Ok for you?

Reply:
Think of this not just as something Spark calls, but a general create service.
The cluster is named, for most clients the group already related to the cluster is correct.
Providing the locations is a special thing where the client can't (or it's more efficient) use the pre-configured configuration.

Spark may always provide the locations, other users of this service may never provide locations.